File: twidere/src/main/java/org/mariotaku/twidere/service/StreamingService.java
package org.mariotaku.twidere.service;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.ContentObserver;
import android.net.Uri;
import android.os.Handler;
import android.os.IBinder;
import android.support.v4.app.NotificationCompat;
import android.support.v4.util.SimpleArrayMap;
import android.text.TextUtils;
import android.util.Log;

import org.mariotaku.microblog.library.MicroBlogException;
import org.mariotaku.microblog.library.twitter.TwitterUserStream;
import org.mariotaku.microblog.library.twitter.UserStreamCallback;
import org.mariotaku.microblog.library.twitter.model.DeletionEvent;
import org.mariotaku.microblog.library.twitter.model.DirectMessage;
import org.mariotaku.microblog.library.twitter.model.Status;
import org.mariotaku.microblog.library.twitter.model.User;
import org.mariotaku.microblog.library.twitter.model.UserList;
import org.mariotaku.microblog.library.twitter.model.Warning;
import org.mariotaku.restfu.http.Authorization;
import org.mariotaku.restfu.http.ContentType;
import org.mariotaku.restfu.http.Endpoint;
import org.mariotaku.restfu.http.HttpResponse;
import org.mariotaku.restfu.http.mime.Body;
import org.mariotaku.sqliteqb.library.Expression;
import org.mariotaku.twidere.BuildConfig;
import org.mariotaku.twidere.Constants;
import org.mariotaku.twidere.R;
import org.mariotaku.twidere.activity.SettingsActivity;
import org.mariotaku.twidere.model.AccountPreferences;
import org.mariotaku.twidere.model.ParcelableAccount;
import org.mariotaku.twidere.model.ParcelableCredentials;
import org.mariotaku.twidere.model.UserKey;
import org.mariotaku.twidere.provider.TwidereDataStore.AccountSupportColumns;
import org.mariotaku.twidere.provider.TwidereDataStore.Accounts;
import org.mariotaku.twidere.provider.TwidereDataStore.Activities;
import org.mariotaku.twidere.provider.TwidereDataStore.DirectMessages;
import org.mariotaku.twidere.provider.TwidereDataStore.Mentions;
import org.mariotaku.twidere.provider.TwidereDataStore.Statuses;
import org.mariotaku.twidere.util.ContentValuesCreator;
import org.mariotaku.twidere.util.DataStoreUtils;
import org.mariotaku.twidere.util.MicroBlogAPIFactory;
import org.mariotaku.twidere.util.TwidereArrayUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.List;

public class StreamingService extends Service implements Constants {

    private static final int NOTIFICATION_SERVICE_STARTED = 1;

    private final SimpleArrayMap<UserKey, UserStreamCallback> mCallbacks = new SimpleArrayMap<>();
    private ContentResolver mResolver;

    private NotificationManager mNotificationManager;

    private UserKey[] mAccountKeys;

    private static final Uri[] MESSAGES_URIS = new Uri[]{DirectMessages.Inbox.CONTENT_URI,
            DirectMessages.Outbox.CONTENT_URI};

    private final ContentObserver mAccountChangeObserver = new ContentObserver(new Handler()) {

        @Override
        public void onChange(final boolean selfChange) {
            onChange(selfChange, null);
        }

        @Override
        public void onChange(final boolean selfChange, final Uri uri) {
            if (!TwidereArrayUtils.contentMatch(mAccountKeys, DataStoreUtils.getActivatedAccountKeys(StreamingService.this))) {
                initStreaming();
            }
        }

    };

    @Override
    public IBinder onBind(final Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        mResolver = getContentResolver();
        mNotificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        if (BuildConfig.DEBUG) {
            Log.d(Constants.LOGTAG, "Stream service started.");
        }
        initStreaming();
        mResolver.registerContentObserver(Accounts.CONTENT_URI, true, mAccountChangeObserver);
    }

    @Override
    public void onDestroy() {
        clearTwitterInstances();
        mResolver.unregisterContentObserver(mAccountChangeObserver);
        if (BuildConfig.DEBUG) {
            Log.d(Constants.LOGTAG, "Stream service stopped.");
        }
        super.onDestroy();
    }

    private void clearTwitterInstances() {
        for (int i = 0, j = mCallbacks.size(); i < j; i++) {
            new Thread(new ShutdownStreamTwitterRunnable(mCallbacks.valueAt(i))).start();
        }
        mCallbacks.clear();
        mNotificationManager.cancel(NOTIFICATION_SERVICE_STARTED);
    }

    @SuppressWarnings("deprecation")
    private void initStreaming() {
        if (!BuildConfig.DEBUG) return;
        setTwitterInstances();
        updateStreamState();
    }

    private boolean setTwitterInstances() {
        final List<ParcelableCredentials> accountsList = DataStoreUtils.getCredentialsList(this, true);
        final UserKey[] accountKeys = new UserKey[accountsList.size()];
        for (int i = 0, j = accountKeys.length; i < j; i++) {
            final ParcelableCredentials credentials = accountsList.get(i);
            accountKeys[i] = credentials.account_key;
        }
        final AccountPreferences[] activatedPreferences = AccountPreferences.getAccountPreferences(this, accountKeys);
        if (BuildConfig.DEBUG) {
            Log.d(Constants.LOGTAG, "Setting up twitter stream instances");
        }
        mAccountKeys = accountKeys;
        clearTwitterInstances();
        boolean result = false;
        for (int i = 0, j = accountsList.size(); i < j; i++) {
            final AccountPreferences preferences = activatedPreferences[i];
            if (!preferences.isStreamingEnabled()) continue;
            final ParcelableCredentials account = accountsList.get(i);
            final Endpoint endpoint = MicroBlogAPIFactory.getEndpoint(account, TwitterUserStream.class);
            final Authorization authorization = MicroBlogAPIFactory.getAuthorization(account);
            final TwitterUserStream twitter = MicroBlogAPIFactory.getInstance(this, endpoint, authorization, TwitterUserStream.class);
            final TwidereUserStreamCallback callback = new TwidereUserStreamCallback(this, account);
            mCallbacks.put(account.account_key, callback);
            new Thread() {
                @Override
                public void run() {
                    twitter.getUserStream(callback);
                    Log.d(Constants.LOGTAG, String.format("Stream %s disconnected", account.account_key));
                    mCallbacks.remove(account.account_key);
                    updateStreamState();
                }
            }.start();
            result |= true;
        }
        return result;
    }

    private void updateStreamState() {
        if (mCallbacks.size() > 0) {
            final Intent intent = new Intent(this, SettingsActivity.class);
            final PendingIntent contentIntent = PendingIntent.getActivity(this, 0, intent,
                    PendingIntent.FLAG_UPDATE_CURRENT);
            final CharSequence contentTitle = getString(R.string.app_name);
            final CharSequence contentText = getString(R.string.timeline_streaming_running);
            final NotificationCompat.Builder builder = new NotificationCompat.Builder(this);
            builder.setOngoing(true);
            builder.setSmallIcon(R.drawable.ic_stat_refresh);
            builder.setContentTitle(contentTitle);
            builder.setContentText(contentText);
            builder.setContentIntent(contentIntent);
            mNotificationManager.notify(NOTIFICATION_SERVICE_STARTED, builder.build());
        } else {
            mNotificationManager.cancel(NOTIFICATION_SERVICE_STARTED);
        }
    }

    static class ShutdownStreamTwitterRunnable implements Runnable {
        private final UserStreamCallback callback;

        ShutdownStreamTwitterRunnable(final UserStreamCallback callback) {
            this.callback = callback;
        }

        @Override
        public void run() {
            if (callback == null) return;
            Log.d(Constants.LOGTAG, "Disconnecting stream");
            callback.disconnect();
        }

    }

    static class TwidereUserStreamCallback extends UserStreamCallback {

        private final Context context;
        private final ParcelableAccount account;
        private final ContentResolver resolver;

        private boolean statusStreamStarted, mentionsStreamStarted;

        public TwidereUserStreamCallback(final Context context, final ParcelableAccount account) {
            this.context = context;
            this.account = account;
            resolver = context.getContentResolver();
        }

        @Override
        public void onConnected() {

        }

        @Override
        public void onBlock(final User source, final User blockedUser) {
            final String message = String.format("%s blocked %s", source.getScreenName(), blockedUser.getScreenName());
            Log.d(LOGTAG, message);
        }

        @Override
        public void onDirectMessageDeleted(final DeletionEvent event) {
            final String where = Expression.equalsArgs(DirectMessages.MESSAGE_ID).getSQL();
            final String[] whereArgs = {event.getId()};
            for (final Uri uri : MESSAGES_URIS) {
                resolver.delete(uri, where, whereArgs);
            }
        }

        @Override
        public void onStatusDeleted(final DeletionEvent event) {
            final String statusId = event.getId();
            resolver.delete(Statuses.CONTENT_URI, Expression.equalsArgs(Statuses.STATUS_ID).getSQL(),
                    new String[]{statusId});
            resolver.delete(Activities.AboutMe.CONTENT_URI, Expression.equalsArgs(Activities.STATUS_ID).getSQL(),
                    new String[]{statusId});
        }

        @Override
        public void onDirectMessage(final DirectMessage directMessage) {
            if (directMessage == null || directMessage.getId() == null) return;
            final String where = Expression.and(Expression.equalsArgs(DirectMessages.ACCOUNT_KEY),
                    Expression.equalsArgs(DirectMessages.MESSAGE_ID)).getSQL();
            final String[] whereArgs = {account.account_key.toString(), directMessage.getId()};
            for (final Uri uri : MESSAGES_URIS) {
                resolver.delete(uri, where, whereArgs);
            }
            final User sender = directMessage.getSender(), recipient = directMessage.getRecipient();
            if (TextUtils.equals(sender.getId(), account.account_key.getId())) {
                final ContentValues values = ContentValuesCreator.createDirectMessage(directMessage,
                        account.account_key, true);
                if (values != null) {
                    resolver.insert(DirectMessages.Outbox.CONTENT_URI, values);
                }
            }
            if (TextUtils.equals(recipient.getId(), account.account_key.getId())) {
                final ContentValues values = ContentValuesCreator.createDirectMessage(directMessage,
                        account.account_key, false);
                final Uri.Builder builder = DirectMessages.Inbox.CONTENT_URI.buildUpon();
                builder.appendQueryParameter(QUERY_PARAM_NOTIFY, "true");
                if (values != null) {
                    resolver.insert(builder.build(), values);
                }
            }

        }

        @Override
        public void onException(final Throwable ex) {
            if (ex instanceof MicroBlogException) {
                Log.w(LOGTAG, String.format("Error %d", ((MicroBlogException) ex).getStatusCode()), ex);
                final HttpResponse response = ((MicroBlogException) ex).getHttpResponse();
                if (response != null) {
                    try {
                        final Body body = response.getBody();
                        if (body != null) {
                            final ByteArrayOutputStream os = new ByteArrayOutputStream();
                            body.writeTo(os);
                            final String charsetName;
                            final ContentType contentType = body.contentType();
                            if (contentType != null) {
                                final Charset charset = contentType.getCharset();
                                if (charset != null) {
                                    charsetName = charset.name();
                                } else {
                                    charsetName = Charset.defaultCharset().name();
                                }
                            } else {
                                charsetName = Charset.defaultCharset().name();
                            }
                            Log.w(LOGTAG, os.toString(charsetName));
                        }
                    } catch (IOException e) {
                        Log.w(LOGTAG, e);
                    }
                }
            } else {
                Log.w(Constants.LOGTAG, ex);
            }
        }

        @Override
        public void onFavorite(final User source, final User target, final Status targetStatus) {
            final String message = String.format("%s favorited %s's tweet: %s", source.getScreenName(),
                    target.getScreenName(), targetStatus.getExtendedText());
            Log.d(LOGTAG, message);
        }

        @Override
        public void onFollow(final User source, final User followedUser) {
            final String message = String
                    .format("%s followed %s", source.getScreenName(), followedUser.getScreenName());
            Log.d(LOGTAG, message);
        }

        @Override
        public void onFriendList(final long[] friendIds) {

        }

        @Override
        public void onScrubGeo(final long userId, final long upToStatusId) {
            final String where = Expression.and(Expression.equalsArgs(Statuses.USER_KEY),
                    Expression.greaterEqualsArgs(Statuses.SORT_ID)).getSQL();
            final String[] whereArgs = {String.valueOf(userId), String.valueOf(upToStatusId)};
            final ContentValues values = new ContentValues();
            values.putNull(Statuses.LOCATION);
            resolver.update(Statuses.CONTENT_URI, values, where, whereArgs);
        }

        @Override
        public void onStallWarning(final Warning warn) {

        }

        @Override
        public void onStatus(final Status status) {
            final ContentValues values = ContentValuesCreator.createStatus(status, account.account_key);
            if (!statusStreamStarted) {
                statusStreamStarted = true;
                values.put(Statuses.IS_GAP, true);
            }
            final String where = Expression.and(Expression.equalsArgs(AccountSupportColumns.ACCOUNT_KEY),
                    Expression.equalsArgs(Statuses.STATUS_ID)).getSQL();
            final String[] whereArgs = {account.account_key.toString(), String.valueOf(status.getId())};
            resolver.delete(Statuses.CONTENT_URI, where, whereArgs);
            resolver.delete(Mentions.CONTENT_URI, where, whereArgs);
            resolver.insert(Statuses.CONTENT_URI, values);
            final Status rt = status.getRetweetedStatus();
            if (rt != null && rt.getExtendedText().contains("@" + account.screen_name) || rt == null
                    && status.getExtendedText().contains("@" + account.screen_name)) {
                resolver.insert(Mentions.CONTENT_URI, values);
            }
        }

        @Override
        public void onTrackLimitationNotice(final int numberOfLimitedStatuses) {

        }

        @Override
        public void onUnblock(final User source, final User unblockedUser) {
            final String message = String.format("%s unblocked %s", source.getScreenName(),
                    unblockedUser.getScreenName());
            Log.d(LOGTAG, message);
        }

        @Override
        public void onUnfavorite(final User source, final User target, final Status targetStatus) {
            final String message = String.format("%s unfavorited %s's tweet: %s", source.getScreenName(),
                    target.getScreenName(), targetStatus.getExtendedText());
            Log.d(LOGTAG, message);
        }

        @Override
        public void onUserListCreation(final User listOwner, final UserList list) {

        }

        @Override
        public void onUserListDeletion(final User listOwner, final UserList list) {

        }

        @Override
        public void onUserListMemberAddition(final User addedMember, final User listOwner, final UserList list) {

        }

        @Override
        public void onUserListMemberDeletion(final User deletedMember, final User listOwner, final UserList list) {

        }

        @Override
        public void onUserListSubscription(final User subscriber, final User listOwner, final UserList list) {

        }

        @Override
        public void onUserListUnsubscription(final User subscriber, final User listOwner, final UserList list) {

        }

        @Override
        public void onUserListUpdate(final User listOwner, final UserList list) {

        }

        @Override
        public void onUserProfileUpdate(final User updatedUser) {

        }
    }

}

File: twidere/src/main/kotlin/org/mariotaku/twidere/service/StreamingService.kt
package org.mariotaku.twidere.service

import android.accounts.AccountManager
import android.accounts.OnAccountsUpdateListener
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import androidx.annotation.UiThread
import androidx.annotation.WorkerThread
import androidx.core.app.NotificationCompat
import androidx.core.net.ConnectivityManagerCompat
import org.mariotaku.abstask.library.TaskStarter
import org.mariotaku.kpreferences.get
import org.mariotaku.ktextension.addOnAccountsUpdatedListenerSafe
import org.mariotaku.ktextension.removeOnAccountsUpdatedListenerSafe
import org.mariotaku.ktextension.toLongOr
import org.mariotaku.ktextension.toNulls
import org.mariotaku.library.objectcursor.ObjectCursor
import org.mariotaku.microblog.library.MicroBlogException
import org.mariotaku.microblog.library.twitter.TwitterUserStream
import org.mariotaku.microblog.library.twitter.annotation.StreamWith
import org.mariotaku.microblog.library.twitter.model.*
import org.mariotaku.sqliteqb.library.Columns
import org.mariotaku.sqliteqb.library.Expression
import org.mariotaku.twidere.R
import org.mariotaku.twidere.TwidereConstants.LOGTAG
import org.mariotaku.twidere.annotation.AccountType
import org.mariotaku.twidere.constant.streamingEnabledKey
import org.mariotaku.twidere.constant.streamingNonMeteredNetworkKey
import org.mariotaku.twidere.constant.streamingPowerSavingKey
import org.mariotaku.twidere.extension.model.*
import org.mariotaku.twidere.extension.model.api.key
import org.mariotaku.twidere.extension.model.api.microblog.toParcelable
import org.mariotaku.twidere.extension.model.api.toParcelable
import org.mariotaku.twidere.extension.queryCount
import org.mariotaku.twidere.model.*
import org.mariotaku.twidere.model.notification.NotificationChannelSpec
import org.mariotaku.twidere.model.pagination.SinceMaxPagination
import org.mariotaku.twidere.model.util.AccountUtils
import org.mariotaku.twidere.provider.TwidereDataStore.*
import org.mariotaku.twidere.task.twitter.GetActivitiesAboutMeTask
import org.mariotaku.twidere.task.twitter.message.GetMessagesTask
import org.mariotaku.twidere.util.DataStoreUtils
import org.mariotaku.twidere.util.DebugLog
import org.mariotaku.twidere.util.IntentUtils
import org.mariotaku.twidere.util.Utils
import org.mariotaku.twidere.util.dagger.DependencyHolder
import org.mariotaku.twidere.util.dagger.GeneralComponent
import org.mariotaku.twidere.util.streaming.TwitterTimelineStreamCallback
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class StreamingService : BaseService() {

    internal lateinit var threadPoolExecutor: ExecutorService
    internal lateinit var handler: Handler

    private val submittedTasks: MutableMap<UserKey, StreamingRunnable<*>> = WeakHashMap()

    private val accountChangeObserver = OnAccountsUpdateListener {
        if (!setupStreaming()) {
            stopSelf()
        }
    }

    override fun onCreate() {
        super.onCreate()
        GeneralComponent.get(this).inject(this)
        threadPoolExecutor = Executors.newCachedThreadPool { runnable ->
            val thread = Thread(runnable)
            thread.priority = Thread.NORM_PRIORITY - 1
            if (runnable is StreamingRunnable<*>) {
                thread.name = "twidere-streaming-${runnable.account.key}"
            }
            return@newCachedThreadPool thread
        }
        handler = Handler(Looper.getMainLooper())
        AccountManager.get(this).addOnAccountsUpdatedListenerSafe(accountChangeObserver, updateImmediately = false)
    }

    override fun onDestroy() {
        submittedTasks.forEach {
            // NOTE: IMPORTANT!!! Before Nougat, forEach { k, v -> } will crash because referenced
            // BiConsumer, which is introduced in Java 8
            val (_, v) = it
            v.cancel()
        }
        threadPoolExecutor.shutdown()
        submittedTasks.clear()
        removeNotification()
        AccountManager.get(this).removeOnAccountsUpdatedListenerSafe(accountChangeObserver)
        super.onDestroy()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (setupStreaming()) {
            return START_STICKY
        }
        stopSelf()
        return START_NOT_STICKY
    }

    override fun onBind(intent: Intent) = throw UnsupportedOperationException()

    /**
     * @return True if there're enabled accounts, false if request not met and service should be stopped
     */
    private fun setupStreaming(): Boolean {
        if (!preferences[streamingEnabledKey]) {
            return false
        }
        if (!activityTracker.isHomeActivityLaunched) {
            return false
        }
        // Quit if no connection
        if (connectivityManager.activeNetworkInfo?.isAvailable != true) {
            return false
        }
        // Quit if connection metered (with preference)
        val isNetworkMetered = ConnectivityManagerCompat.isActiveNetworkMetered(connectivityManager)
        if (preferences[streamingNonMeteredNetworkKey] && isNetworkMetered) {
            return false
        }
        // Quit if not charging (with preference)
        val isCharging = Utils.isCharging(this)
        if (preferences[streamingPowerSavingKey] && !isCharging) {
            return false
        }
        // Quit if no streaming instance available
        if (!updateStreamingInstances()) {
            return false
        }
        showNotification()
        return true
    }

    private fun updateStreamingInstances(): Boolean {
        val am = AccountManager.get(this)
        val supportedAccounts = AccountUtils.getAllAccountDetails(am, true).filter { it.isStreamingSupported }
        val supportedPrefs = supportedAccounts.map { AccountPreferences(this, preferences, it.key) }
        val enabledAccounts = supportedAccounts.filter { account ->
            return@filter supportedPrefs.any {
                account.key == it.accountKey && it.isStreamingEnabled
            }
        }

        if (enabledAccounts.isEmpty()) return false

        // Remove all disabled instances
        submittedTasks.forEach {
            // NOTE: IMPORTANT!!! Before Nougat, forEach { k, v -> } will crash because referenced
            // BiConsumer, which is introduced in Java 8
            val (k, v) = it
            if (enabledAccounts.none { k == it.key } && !v.cancelled) {
                v.cancel()
            }
        }
        // Add instances if not running
        enabledAccounts.forEach { account ->
            val existing = submittedTasks[account.key]
            if (existing == null || existing.cancelled) {
                val runnable = newStreamingRunnable(account, supportedPrefs.first {
                    it.accountKey == account.key
                }) ?: return@forEach
                threadPoolExecutor.submit(runnable)
                submittedTasks[account.key] = runnable
            }
        }
        return true
    }

    private fun showNotification() {
        val intent = IntentUtils.settings("streaming")
        val contentIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)
        val contentTitle = getString(R.string.app_name)
        val contentText = getString(R.string.timeline_streaming_running)
        val builder = NotificationChannelSpec.serviceStatuses.notificationBuilder(this)
        builder.setOngoing(true)
        builder.setSmallIcon(R.drawable.ic_stat_streaming)
        builder.setContentTitle(contentTitle)
        builder.setContentText(contentText)
        builder.setContentIntent(contentIntent)
        builder.setCategory(NotificationCompat.CATEGORY_STATUS)
        builder.priority = NotificationCompat.PRIORITY_MIN
        startForeground(NOTIFICATION_SERVICE_STARTED, builder.build())
    }

    private fun removeNotification() {
        stopForeground(true)
    }

    private fun newStreamingRunnable(account: AccountDetails, preferences: AccountPreferences): StreamingRunnable<*>? {
        when (account.type) {
            AccountType.TWITTER -> {
                return TwitterStreamingRunnable(this, account, preferences)
            }
        }
        return null
    }

    internal abstract class StreamingRunnable<T>(
            val context: Context,
            val account: AccountDetails,
            val accountPreferences: AccountPreferences
    ) : Runnable {

        var cancelled: Boolean = false
            private set

        override fun run() {
            val instance = createStreamingInstance()
            while (!cancelled && !Thread.currentThread().isInterrupted) {
                try {
                    instance.beginStreaming()
                } catch (e: MicroBlogException) {
                    DebugLog.w(LOGTAG, msg = "Can't stream for ${account.key}", tr = e)
                }
                Thread.sleep(TimeUnit.MINUTES.toMillis(1))
            }
        }

        fun cancel(): Boolean {
            if (cancelled) return false
            cancelled = true
            onCancelled()
            return true
        }

        abstract fun createStreamingInstance(): T

        abstract fun T.beginStreaming()

        abstract fun onCancelled()
    }

    internal inner class TwitterStreamingRunnable(
            context: Context,
            account: AccountDetails,
            accountPreferences: AccountPreferences
    ) : StreamingRunnable<TwitterUserStream>(context, account, accountPreferences) {

        private val profileImageSize = context.getString(R.string.profile_image_size)
        private val isOfficial = account.isOfficial(context)

        private var canGetInteractions: Boolean = true
        private var canGetMessages: Boolean = true

        private val interactionsTimeoutRunnable = Runnable {
            canGetInteractions = true
        }

        private val messagesTimeoutRunnable = Runnable {
            canGetMessages = true
        }

        val callback = object : TwitterTimelineStreamCallback(account.key.id) {

            private var lastStatusTimestamps = LongArray(2)

            private var homeInsertGap = false
            private var interactionsInsertGap = false

            private var lastActivityAboutMe: ParcelableActivity? = null

            override fun onConnected(): Boolean {
                homeInsertGap = true
                interactionsInsertGap = true
                return true
            }

            override fun onHomeTimeline(status: Status): Boolean {
                if (!accountPreferences.isStreamHomeTimelineEnabled) {
                    homeInsertGap = true
                    return false
                }
                val parcelableStatus = status.toParcelable(account, profileImageSize = profileImageSize)
                parcelableStatus.is_gap = homeInsertGap

                val currentTimeMillis = System.currentTimeMillis()
                if (lastStatusTimestamps[0] >= parcelableStatus.timestamp) {
                    val extraValue = (currentTimeMillis - lastStatusTimestamps[1]).coerceAtMost(499)
                    parcelableStatus.position_key = parcelableStatus.timestamp + extraValue
                } else {
                    parcelableStatus.position_key = parcelableStatus.timestamp
                }
                parcelableStatus.inserted_date = currentTimeMillis

                lastStatusTimestamps[0] = parcelableStatus.position_key
                lastStatusTimestamps[1] = parcelableStatus.inserted_date

                val values = ObjectCursor.valuesCreatorFrom(ParcelableStatus::class.java)
                        .create(parcelableStatus)
                context.contentResolver.insert(Statuses.CONTENT_URI, values)
                homeInsertGap = false
                return true
            }

            override fun onActivityAboutMe(activity: Activity): Boolean {
                if (!accountPreferences.isStreamInteractionsEnabled) {
                    interactionsInsertGap = true
                    return false
                }
                if (isOfficial) {
                    // Wait for 30 seconds to avoid rate limit
                    if (canGetInteractions) {
                        handler.post { getInteractions() }
                        canGetInteractions = false
                        handler.postDelayed(interactionsTimeoutRunnable, TimeUnit.SECONDS.toMillis(30))
                    }
                } else {
                    val insertGap: Boolean
                    if (activity.action in Activity.Action.MENTION_ACTIONS) {
                        insertGap = interactionsInsertGap
                        interactionsInsertGap = false
                    } else {
                        insertGap = false
                    }
                    val curActivity = activity.toParcelable(account, insertGap, profileImageSize)
                    curActivity.account_color = account.color
                    curActivity.position_key = curActivity.timestamp
                    var updateId = -1L
                    if (curActivity.action !in Activity.Action.MENTION_ACTIONS) {
                        /* Merge two activities if:
                         * * Not mention/reply/quote
                         * * Same action
                         * * Same source or target or target object
                         */
                        val lastActivity = this.lastActivityAboutMe
                        if (lastActivity != null && curActivity.action == lastActivity.action) {
                            if (curActivity.reachedCountLimit) {
                                // Skip if more than 10 sources/targets/target_objects
                            } else if (curActivity.isSameSources(lastActivity)) {
                                curActivity.prependTargets(lastActivity)
                                curActivity.prependTargetObjects(lastActivity)
                                updateId = lastActivity._id
                            } else if (curActivity.isSameTarget(lastActivity)) {
                                curActivity.prependSources(lastActivity)
                                curActivity.prependTargets(lastActivity)
                                updateId = lastActivity._id
                            } else if (curActivity.isSameTargetObject(lastActivity)) {
                                curActivity.prependSources(lastActivity)
                                curActivity.prependTargets(lastActivity)
                                updateId = lastActivity._id
                            }
                            if (updateId > 0) {
                                curActivity.min_position = lastActivity.min_position
                                curActivity.min_sort_position = lastActivity.min_sort_position
                            }
                        }
                    }
                    val values = ObjectCursor.valuesCreatorFrom(ParcelableActivity::class.java)
                            .create(curActivity)
                    val resolver = context.contentResolver
                    if (updateId > 0) {
                        val where = Expression.equals(Activities._ID, updateId).sql
                        resolver.update(Activities.AboutMe.CONTENT_URI, values, where, null)
                        curActivity._id = updateId
                    } else {
                        val uri = resolver.insert(Activities.AboutMe.CONTENT_URI, values)
                        if (uri != null) {
                            curActivity._id = uri.lastPathSegment.toLongOr(-1L)
                        }
                    }
                    lastActivityAboutMe = curActivity
                }
                return true
            }

            @WorkerThread
            override fun onDirectMessage(directMessage: DirectMessage): Boolean {
                if (!accountPreferences.isStreamDirectMessagesEnabled) {
                    return false
                }
                if (canGetMessages) {
                    handler.post { getMessages() }
                    canGetMessages = false
                    val timeout = TimeUnit.SECONDS.toMillis(if (isOfficial) 30 else 90)
                    handler.postDelayed(messagesTimeoutRunnable, timeout)
                }
                return true
            }

            override fun onAllStatus(status: Status) {
                if (!accountPreferences.isStreamNotificationUsersEnabled) {
                    return
                }
                val user = status.user ?: return
                val userKey = user.key
                val where = Expression.and(Expression.equalsArgs(CachedRelationships.ACCOUNT_KEY),
                        Expression.equalsArgs(CachedRelationships.USER_KEY),
                        Expression.equals(CachedRelationships.NOTIFICATIONS_ENABLED, 1)).sql
                val whereArgs = arrayOf(account.key.toString(), userKey.toString())
                if (context.contentResolver.queryCount(CachedRelationships.CONTENT_URI,
                        where, whereArgs) <= 0) return

                contentNotificationManager.showUserNotification(account.key, status, userKey)
            }

            override fun onStatusDeleted(event: DeletionEvent): Boolean {
                val deleteWhere = Expression.and(Expression.likeRaw(Columns.Column(Statuses.ACCOUNT_KEY), "'%@'||?"),
                        Expression.equalsArgs(Columns.Column(Statuses.ID))).sql
                val deleteWhereArgs = arrayOf(account.key.host, event.id)
                context.contentResolver.delete(Statuses.CONTENT_URI, deleteWhere, deleteWhereArgs)
                return true
            }

            override fun onDisconnectNotice(code: Int, reason: String?): Boolean {
                disconnect()
                return true
            }

            override fun onException(ex: Throwable): Boolean {
                DebugLog.w(LOGTAG, msg = "Exception for ${account.key}", tr = ex)
                return true
            }

            override fun onUnhandledEvent(obj: TwitterStreamObject, json: String) {
                DebugLog.d(LOGTAG, msg = "Unhandled event ${obj.determine()} for ${account.key}: $json")
            }

            @UiThread
            private fun getInteractions() {
                val task = GetActivitiesAboutMeTask(context)
                task.params = object : RefreshTaskParam {
                    override val accountKeys: Array<UserKey> = arrayOf(account.key)

                    override val pagination by lazy {
                        val keys = accountKeys.toNulls()
                        val sinceIds = DataStoreUtils.getRefreshNewestActivityMaxPositions(context,
                                Activities.AboutMe.CONTENT_URI, keys)
                        val sinceSortIds = DataStoreUtils.getRefreshNewestActivityMaxSortPositions(context,
                                Activities.AboutMe.CONTENT_URI, keys)
                        return@lazy Array(keys.size) { idx ->
                            SinceMaxPagination.sinceId(sinceIds[idx], sinceSortIds[idx])
                        }
                    }

                }
                TaskStarter.execute(task)
            }

            @UiThread
            private fun getMessages() {
                val task = GetMessagesTask(context)
                task.params = object : GetMessagesTask.RefreshMessagesTaskParam(context) {
                    override val accountKeys: Array<UserKey> = arrayOf(account.key)
                }
                TaskStarter.execute(task)
            }
        }

        override fun createStreamingInstance(): TwitterUserStream {
            return account.newMicroBlogInstance(context, cls = TwitterUserStream::class.java)
        }

        override fun TwitterUserStream.beginStreaming() {
            getUserStream(StreamWith.USER, callback)
        }

        override fun onCancelled() {
            callback.disconnect()
        }

    }

    companion object {

        private val NOTIFICATION_SERVICE_STARTED = 1

        fun startOrStopService(context: Context) {
            val streamingIntent = Intent(context, StreamingService::class.java)
            val holder = DependencyHolder.get(context)
            try {
                if (holder.activityTracker.isHomeActivityLaunched) {
                    context.startService(streamingIntent)
                } else {
                    context.stopService(streamingIntent)
                }
            } catch (e: IllegalStateException) {
                // This shouldn't happen, catch it.
            }
        }
    }

}

